feat: add waiig book review (#133)
jsjoeio/joeprevite.com: content/2020-05-22-book-review-writing-an-interpreter-in-go/index.md

---
slug: 'book-review-writing-an-interpreter-in-go'
date: '2020-05-22'
title: 'Book Review: Writing an Interpreter in Go'
description: 'My notes and thoughts on Writing an Interpreter in Go by Thorsten Ball'
tagline: 'My notes and thoughts'
published: true
---

![Writing an interpreter in go paperback on wood](./book_cover.jpg)

When I started my new job, I was told that I would be the "Programming Languages" person. I knew I wanted to learn more not only about specific programming languages like Rust and Reason, but understand them on a higher level.

> I believe it was [@sugarpirate](https://twitter.com/sugarpirate_) who originally recommended both this book and the sequel "[Writing a Compiler in Go](https://compilerbook.com/)" by Thorsten Ball. I was hesitant to purchase the book because I had never written Go before and wasn't planning to. However, others insisted that you didn't need to know Go to read this book.

Reviewer: Just a suggestions 😄 

```suggestion
I believe it was [@sugarpirate](https://twitter.com/sugarpirate_) who originally recommended both [this book](https://interpreterbook.com) and the sequel "[Writing a Compiler in Go](https://compilerbook.com/)" by Thorsten Ball. I was hesitant to purchase the book because I had never written Go before and wasn't planning to. However, others insisted that you didn't need to know Go to read this book.
```

    - [eight infix operators](#eight-infix-operators)
    - [comparing values directly](#comparing-values-directly)
    - [return statements](#return-statements)
    - [bindings](#bindings)
    - [environment](#environment)
    - [first-class citizens](#first-class-citizens)
  - [Chapter 4 - Extending the Interpreter](#chapter-4---extending-the-interpreter)
    - [built-in functions](#built-in-functions)
    - [validate hash keys in evaluation stage](#validate-hash-keys-in-evaluation-stage)
  - [Chapter 5 - Macro System](#chapter-5---macro-system)
- [Overall Thoughts](#overall-thoughts)
- [Recommend the Book?](#recommend-the-book)
- [Purchase the Book](#purchase-the-book)

## Notes

This is a collection of my notes from each chapter in the book, along with my overall thoughts and recommendations.

### Introduction

In the beginning of the book, Thorsten explains why he wrote this book and who it's for. One thing that stuck out to me was this,

> So I wrote this book, for you and me. This is the book I wish I had. This is a book for people who love to look under the hood. For people that love to learn by understanding how something really works.
>
> -Ball 7

> The part that resonated with me was the last sentence in hat quote -- "people that love to learn by understanding how something really works." That made me glad I picked this book up.

Reviewer: ```suggestion
The part that resonated with me was the last sentence in that quote -- "people that love to learn by understanding how something really works." That made me glad I picked this book up.
```

> "Tree-walking interpreter"

> I highlighted this because I know I'll forgot what it means if I don't write it down. Summarized, it means an interpreter that parses the source code, builds an abstract syntax tree (AST) and then evaluates it. The "tree" comes from AST. The "walking" comes from walking down the tree.

Reviewer: ```suggestion
I highlighted this because I know I'll forget what it means if I don't write it down. Summarized, it means an interpreter that parses the source code, builds an abstract syntax tree (AST) and then evaluates it. The "tree" comes from AST. The "walking" comes from walking down the tree.
```

Lexer vs. parser?

> I felt confused in the beginning what the difference once. In my own words, the lexer takes input like characters "1;" and goes, "Ah! Yes. That input is equal to these tokens {token.INTEGER, '1', token.SEMICOLON, ';'}" The parser says, "Ah! Thank you lexer for the tokens, I will turn those into an AST."

Reviewer: ```suggestion
I felt confused in the beginning as to what the difference was. In my own words, the lexer takes input like characters "1;" and goes, "Ah! Yes. That input is equal to these tokens {token.INTEGER, '1', token.SEMICOLON, ';'}" The parser says, "Ah! Thank you lexer for the tokens, I will turn those into an AST."
```

> Every interpreter is built to interpret a specific programming language. That's how you "implement" a programming language. Without a compiler or an interpreter, a programming language is nothing more than an idea or a specification.
>
> -Ball 8

This was an aha moment for me. The compiler and the interpreter are key. I had no idea. I had been writing JavaScript for so long and never really looked under the hood to understand how the interpreter worked. I'm also learning Rust, which is my first compiled language, so hearing this helped my understanding there as well.

#### hashes vs hashmaps vs dictionaries vs objects

There is probably more to it than I know, but I find that this data structure has a different name depending on which language in which you're writing it:

```javascript
let previte = { name: 'Joe', age: 26 }
```

> In the book (which uses Go), the author refers to it as a hash. In JavaScript, I would call this an object. In Python, one might call it a dictionary. Why can't we all decide on one name and use it across all languages? If someone knows the answer to this, please fill me in!

Reviewer: Imagine me waving my hand here, slightly moving my head and saying "Hmm, well, it's not that easy..."

What Python calls dictionaries is what Go calls maps and what's usually referred to as hash table (hash for short, or hash map): https://en.wikipedia.org/wiki/Hash_table

JavaScript objects are their own datatype (remember, they also have methods and attributes, which hash tables do _not_ have). Under the hood, a JavaScript object can be implemented with hash tables, though. But that's just an implementation detail when we're talking about whether JavaScript objects are hash tables: they're not.

But what gets everyone confused is that their _literal_ syntax looks really similar (`map[string]string{"foobar": "barfoo"}` in Go and `{"foobar": "barfoo"}` in JavaScript) and that hash tables in other languages often encode to objects in JSON.

Author: 🤯 OH! That makes a lot more sense - I'm going to include this in the post and quote you! (as long as that's cool!)

> I'd bet that you can follow this book along even if you've never written a single line of Go in your life.
>
> -Ball 10

I tripped myself up in the beginning when initializing a Go project locally, but quickly solved that thanks to his help and friends on Twitter. So he was right! I don't think the language was an obstacle in anyway.

> And if you, after reading this book, choose to write your own interpreter in another language this should come in handy.

I wrote a note to remember this! I don't know if I'll do this, but it could be helpful for learning Reason or Rust more deeply. I'll leave this here as a reminder.

### Chapter 1 - Lexing

The first part of the interpreter: lexical analysis.

#### lexical analysis, lexing, and lexer

Lots of words that sound and look similar, but what do they all mean? "Lexical analysis" is taking the source code and transforming it into tokens. This process is also called "lexing." The "lexer" is the thing that does this. Thorsten mentions that it's also sometimes called tokenizer or scanner.

https://excalidraw.com/#json=5086916491870208,sLUFfW5oMgrO3ApQwe4YQw

> ...we're going to lex in our first step
>
> -Ball 15

> I wrote this down to highlight **lex** and remind myself how to use it in a sentence. An example, "The first thing our interpreter does is lex the source code to generate tokens. The tokens are then parsed into an abstract syntax tree."

Reviewer: I'm honestly not sure whether using "lex" as a verb is correct :) What others are also using is "tokenize" or "scan".

Author: Whoops! Let me add a note here

#### identifiers

These are the variable names. It makes sense too. They are identified by these variable names. They should also be identified by our interpreter and our program.

https://excalidraw.com/#json=5643057177296896,5i_HmKWn3lDRcBRDh9UMdw

#### token types

One of my favorite parts was learning about the limited token types. Thorsten had us create these token types as constants and each one could be used for different characters in our programming language. Here are some examples:

```go
const (

> 	EOF = "EOF" // end of file

Reviewer: ```suggestion
  EOF = "EOF" // end of file
```

I don't think I had heard this term/concept mentioned before, but he was referring to JSON, YAML, TOML, INI, etc.

#### syntactic analysis

Again, I don't think I had heard this before.

> ...the process of parsing is also call syntactic analysis
>
> -Ball 39

#### parser generator

Another new term, Thorsten explains that parser generators are "tools that, when fed with a formal description of a language, produce parsers as their output" (Ball 39).

To be honest, this is still a bit foreign to me. Of course when I google it, the term "[compiler-compiler](https://en.wikipedia.org/wiki/Compiler-compiler)" comes up which doesn't help. I think I'll leave it at that for now, but may revisit.

> Parsing is one of he most well-understood branches of computer science...
>
> -Ball 40

#### top-down vs bottom-up vs recursive decent parsing

I bet I could write a blog post on this alone (I don't know all of the differences, and won't explore them here). The method we take in the book is recursive decent parsing which he mentions is a "top down operator precedence" parser and also called "Pratt parser" after Vaughan Pratt.

#### binding vs identifier vs expression

> I wrote a note to myself "define binding identifier vs. expression." Let's give it a shot. Let statements like `let x = 5;` bind the expression "5" to the identifier "x". And as Thorsten reminds us, "Expressions produce values, statments don't" (Ball 42). To be clear, if we take this `let x = 5;` it is a statement because it does not produce a value. However, this `5` does produce a value. It's like a tupperware container with some food inside. The tupperware doesn't give us food, but if we go inside and look, we get food.

Reviewer: There's an `e` missing in `statement`:

```suggestion
I wrote a note to myself "define binding identifier vs. expression." Let's give it a shot. Let statements like `let x = 5;` bind the expression "5" to the identifier "x". And as Thorsten reminds us, "Expressions produce values, statements don't" (Ball 42). To be clear, if we take this `let x = 5;` it is a statement because it does not produce a value. However, this `5` does produce a value. It's like a tupperware container with some food inside. The tupperware doesn't give us food, but if we go inside and look, we get food.
```

As a reminder to myself, `prefixParseFns` is used in the book and is the same as "nuds" for "null denotations" by Pratt. `infixParseFns` is equivalent to "leds" or "left denotations" by Pratt.

#### terminology

Thorsten points out some useful terms on page 63:

- **prefix operator** - in front of operand
- **postfix operator** - after operand like `foobar++`
- **infix operator** - in between operands
- **binary expressions** - the operator has two operands like `5 * 8`
- **operator precedence** - also called "order of operations" (ring a bell?)

#### IntegerLiteral

I wrote a note to show an example of what an integer literal data structure looks like in our language:

```go
type IntegerLiteral struct {
  Token token.Token
  Value int64
}
```

As you can see, we store the token type and the actual value. For example, `5` would look like this:

```go
{
  Token: token.INT, // integer
  Value: "5"
}
```

Pretty cool, huh?

Sidenote: I love that he abbreviates literal to `lit` in our helper functions. I laugh because it reminds me of the colloquial saying, "that's so [lit](https://www.urbandictionary.com/define.php?term=lit)!"

#### prefixParseFns and infixParseFns

> I found this part of the book a bit confusing. I'm struggling to even explain what I don't know...If I have the integer `5`, why do I call `prefixParseFns`? Is it because I need to check if `5` has a prefix such as a minus sign "-"? Ah...now that I'm writing this, I think that's it. Because as a human, I can look at `-5` and say, "Yes, that has a prefix we need to parse." But as a machine, I don' know that from looking at the token alone because these tokens `5` and `-5` are the same token type but have different values.

Reviewer: You got it. The parser always needs to start parsing what he expects to be there. That's why where start calling `parseStatement` when kicking off the parser, since that _should_ be in the source code.

There's a `t` missing in `don't`:

```suggestion
I found this part of the book a bit confusing. I'm struggling to even explain what I don't know...If I have the integer `5`, why do I call `prefixParseFns`? Is it because I need to check if `5` has a prefix such as a minus sign "-"? Ah...now that I'm writing this, I think that's it. Because as a human, I can look at `-5` and say, "Yes, that has a prefix we need to parse." But as a machine, I don't know that from looking at the token alone because these tokens `5` and `-5` are the same token type but have different values.
```

Author: Ahhh! Okay, that makes sense! Thanks for validating that!

```go
func (p *Parser) parseGroupedExpression() ast.Expression {
  p.nextToken()
  exp := p.parseExpression(LOWEST) // this LOWEST thing I don't get

  if !p.expectPeek(token.RPAREN) {
    return nil
  }

  return exp
}
```

#### block statements

These are a series of statements which are surrounded by `{` and `}`.

#### infix parse functions

I wrote a note later where he says,

> Yes, we need to register an `infixParseFn` for `token.LPAREN`. This way we parse the expression that is the function (either an identifier, or a function literal), then check for an `infixParseFn` associated with `token.LPAREN` and call it with the already parsed expression as argument. And in this `infixParseFn` we can then parse the argument list.
>
> -Ball 122

I'm blocking myself here. In my mind, I think "infix === things like +, -, \*, /" so it feels strange here to use an LPAREN or "(" as an infix, but I guess infix can also mean in between two other characters. Example: `add(5, 5)` where the `(` is in between the "d" and the "5". Still, future me should revisit this.

#### serialization

Thorsten mentions this,

> What we want is an AST that (serialized as a string)...
>
> -Ball 89

> And I thought, "Do I really know what this means?" And the answer is no, no I do not. Wikipedia [explains it](https://en.wikipedia.org/wiki/Serialization) well. I'll summarize it as taking something, translating it into a data structure or object and storing it for later use (possibly restructuring it).

Reviewer: The colloquial term would be "dumping" 😉 As in "What we want is an AST that (dumped as a string)..."

Or "converting".

Author: ahhh! converting helps! 😄

### Chapter 3 - Evaluation

This is the part where the abstract syntax trees are turned into something meaningful.

#### interpreters vs compilers

> The notion of an interpreter as something that doesn't leave executable artifacts behind (in contract to a compiler, which does just that) gets fuzzy real fast when looking at the implementations of real-world and highly-optimized programming languages.
>
> -Ball 132

I found this super helpful (at least comparing it to say Rust). Interpreter has one-step. Compilers have two.

#### tree-walking interpreters

> I wrote down this because I wanted to look up a few examples of programming languages that fall into this category. I don't know how trustworthy [these Reddit comments](https://www.reddit.com/r/ProgrammingLanguages/comments/dfi9lk/viability_of_treewalk_interpreters/) are but here are a few notes:

Reviewer: ```suggestion
I wrote this down because I wanted to look up a few examples of programming languages that fall into this category. I don't know how trustworthy [these Reddit comments](https://www.reddit.com/r/ProgrammingLanguages/comments/dfi9lk/viability_of_treewalk_interpreters/) are but here are a few notes:
```

> Ruby was a tree-walk interpreter before version 1.9 (I think) and was used in production quite a bit, albeit with a reputation for being very slow and consuming a lot of RAM.

> R was also a tree-walking interpreter for around 20 years! It only changed in the last 5 or so years, a similar time as Ruby.

> It seems Perl still is a tree-walk interpreter.

A bit later, Thorsten does touch on a few. He mentions the Ruby example.

#### intermediate representation

I hadn't heard this term before. According to [Wikipedia](https://en.wikipedia.org/wiki/Intermediate_representation),

> > intermediate representation is the [data structure](https://en.wikipedia.org/wiki/Data_structure) or code used internally by a [compiler](https://en.wikipedia.org/wiki/Compiler) or [virtual machine](https://en.wikipedia.org/wiki/Virtual_machine) to represent [source code](https://en.wikipedia.org/wiki/Source_code).

Reviewer: FWIW, and I don't know if I mention this in the book or not, Intermediate Representation is often shortened to "IR". LLVM IR, for example, is the intermediate representation used by the LLVM compiler toolchain. That often pops up when there's talk around the Rust compiler, for example, which uses LLVM :)

Author: I'm sure you do! Ahhh, okay. That makes sense :) I heard LLVM, but it's never something I've understood. Thanks for mentioning that!

#### just in time

He also mentions on page 133 a JIT, or "just in time" interpreter/compiler. From what I can tell, [JavaScript](https://blog.sessionstack.com/how-javascript-works-inside-the-v8-engine-5-tips-on-how-to-write-optimized-code-ac089e62b12e) usually uses this technique.

One thing to note as well,

> An interpreter that compiles to byte code and uses a virtual machine to evaluate said bytecode is going to be a lot faster.
>
> -Ball 133

This is compared to a "tree-walking interpreter that recursively evaluates an AST."

He also mentions WebKit JavaScript engine JavaScriptCore and the interpreter "SquirrelFish" (weird name, right?). This one has "four different stages of JIT compilation."

Lua is another languages that adopted a JIT, but 12 years after the first release (Ball 134).

#### host language

It was interesting that for our evaluator, Thorsten said we only needed two things,

> a tree-walking evaluator and a way to represent Monkey values in our host language Go.
>
> -Ball 135

> This stuck out to me because I realized Go, the language in which we were writing the language, is this said host language. It made me think of Reason, whose host language is OCaml. But I believe when you look at Rust, it's host language is Rust. And same with TypeScript whose host language is TypeScript. Weird, but cool!

Reviewer: That's why they're called self-hosted compilers 🙂

Small typo:

```suggestion
This stuck out to me because I realized Go, the language in which we were writing the language, is this said host language. It made me think of Reason, whose host language is OCaml. But I believe when you look at Rust, its host language is Rust. And same with TypeScript whose host language is TypeScript. Weird, but cool!
```

Author: 🤯 I had no idea! Adding that in here as a side note

In this chapter, the first one we implement is `quote`. As explained,

> it stops its argument from being evaluated. Instead it returns the AST node representing the argument.
>
> -Ball (5.3 - Quote)

This was hard for me to grasp. I couldn't think of a practical use case where I would want to do this. After I spend more time understanding macro systems and have a few under my belt, this will make more sense.

Some other highlights:

> Conceptually, “macro expansion” means evaluating all calls to macros in the source code and replacing them with the return value of this evaluation. Macros take source code as their input and return source code, so by calling them we “expand” the source code, because each call might result in more of it.
>
> -Ball (5.5 - Macro Expansion)

> Take a look at Elixir or any Lisp, for inspiration, to see how the syntax gives power to the macro systems and how that in turn makes the language itself more powerful and expressive
>
> -Ball (5.7 - Dream On… In Macros)

## Overall Thoughts

> I really enjoyed the book! I feel like I unlocked a new understand. Now, I know how interpreters work. And thanks to that, I have a better understanding of what's happening under the hood! Following along with the author felt straightforward.

Reviewer: ```suggestion
I really enjoyed the book! I feel like I unlocked a new understanding. Now, I know how interpreters work. And thanks to that, I have a better understanding of what's happening under the hood! Following along with the author felt straightforward.
```

I now have a [fully-functioning interpreter](https://github.com/jsjoeio/monkey)! I never thought I would do something like this, and I probably wouldn't have attempted it on my own without this book.

## Recommend the Book?

If you're interested in deepening your understanding of programming languages, I recommend this book! It's practical. It's funny. He walks you through everything and breaks concepts down so they're easy to understand.

It's fantastic way to learn what it takes to write your own language. And at the end of it, you'll have new knowledge, and something you can show your friends.

## Purchase the Book

> If you want to purchase the book, you can do so here at [interpreterbook.com](https://interpreterbook.com/). You can even buy this book plus his [compiler book](https://compilerbook.com/) in a [bundle](https://gumroad.com/l/waiig_wacig_bundle) and save! Highly recommend.

Reviewer: Thanks for the recommendation! :)

Author: Thanks for writing an awesome book! 😄
